text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+embedded manifest.
+There is no required way to distinguish a single-file
+`.rs` package from any other `.rs` file.
+
+A single-file package may contain an embedded manifest.
+An embedded manifest is stored using `TOML` in rust "frontmatter",
+a markdown code-fence with `cargo` at the start of the infostring at the top of
+the file.
+
+Inferred / defaulted manifest fields:
+- `package.name = <slugified file stem>`
+- `package.version = "0.0.0"` to [call attention to this crate being used in unexpected places](https://matklad.github.io/2021/08/22/large-rust-workspaces.html#Smaller-Tips)
+  - This may be altered with [rust-lang/cargo#9829](https://github.com/rust-lang/cargo/issues/9829)
+- `package.publish = false` to avoid accidental publishes, particularly if we
+  later add support for including them in a workspace.
+- `package.edition = <current>` to avoid always having to add an embedded
+  manifest at the cost of potentially breaking scripts on rust upgrades
+  - Warn when `edition` is unspecified.  Since piped commands are run with `--quiet`, this may not show up.
+  - Based on feedback, we might add `cargo-<edition>-edition` proxies to put in `#!` as a shorthand
+  - Based on feedback, we can switch to "edition is required as of <future> edition"
+
+Disallowed manifest fields:
+- `[workspace]`, `[lib]`, `[[bin]]`, `[[example]]`, `[[test]]`, `[[bench]]`
+- `package.workspace`, `package.build`, `package.links`, `package.publish`, `package.autobins`, `package.autoexamples`, `package.autotests`, `package.autobenches`
+
+Single-file packages maintain an out-of-line target directory by default.
+This is `$CARGO_HOME/target/<hash-of-path>`.
+
+A single-file package is accepted by cargo commands as a `--manifest-path`
+- Files are considered to have embedded manifest if they end with `.rs` or they lack and extension and are of type file
+- This allows running `cargo test --manifest-path single.rs`
+- `cargo add` and `cargo remove` may not support editing embedded manifests initially
+- Path-dependencies may not refer to single-file packages at this time (they don't have a `lib` target anyways)
+
+Single-file packages will not be accepted as `path` or `git` dependencies.
+
+The lockfile for single-file packages will be placed in `CARGO_TARGET_DIR`.  In
+the future, when workspaces are supported, that will allow a user to have a
+persistent lockfile.
+We may also allow customizing the non-workspace lockfile location in the [future](#future-possibilities).
+
+## `cargo <file>.rs`
+
+`cargo` is intended for putting in the `#!` for single-file packages:
+```rust
+#!/usr/bin/env cargo
+
+fn main() {
+    println!("Hello world");
+}
+```
+- Like with `cargo install`, `.cargo/config.toml` will be read based on the
+  scripts location rather than the current-dir.
+  - And like `cargo install`, the current-dir rustup-toolchain is respected
+- `--release` is not passed in because the primary use case is for exploratory
```

Comment:
Some ideas for how this situation can be improved:
* To address this in the future, there could be a manifest field that allows specifying a default profile used by `cargo run` in place of the `dev` profile. (This is orthogonal to single-file packages.)
* Today, a single-file-package author will be able to alter the `dev` profile to be optimized or otherwise similar to `release`, if they consider it a priority.

Reply:
I think this is worth considering carefully. People will use this for quick scripts, but the fact that they're writing the scripts in *rust* may be for (among many other reasons) performance. So we should have a *very simple* way to switch to release mode.

Reply:
This is at least called out now in f78441ff and 9002faf4.f we do add a `run` profile, just doing

If we add the suggested `run` profile, we could ensure the following works
```toml
[profile.run]
inherits = "release"
```
That seems like a relatively light way to switch it

Reply:
It would be nice if we could read arguments from the shebang, e.g.

```
#!/usr/bin/env cargo --release
```

but I imagine there are a good number of cross-platform wrinkles here.

Reply:
Yes, one of our items under "Naming: Considerations" is

> The command that is run in a `#!` line should not require arguments (e.g. not
> `#!/usr/bin/env cargo <something>`) because it will fail.  `env` treats the
> rest of the line as the bin name, spaces included.  You need to use `env -S`
> but that wasn't supported on macOS at least, last I tested.

Reply:
Just noticed we have some weird profile special cases.  I wonder if it'd be acceptable to add one more: if `run` does not have a defined table, you get the behavior of `dev`, including the directory name.

Reply:
We talked about this in the cargo team meeting.  We felt that there are viable, non-breaking ways of addressing this later so we decided to defer address this as it isn't the primary focus of this first phase of this work.

This is covered in https://github.com/epage/rfcs/blob/script/text/3502-cargo-script.md#access-to---release-in-shebang